src/storagetm/storagetm.c

```diff
@@ -0,0 +1,1064 @@
+                         IN_ADDR_TO_STRING(a->family, &a->address),
+                         port->portnr);
+        }
+
+        return 0;
+}
+
+static int nvme_port_report(NvmePort *port) {
+        assert(port);
+
+        if (!port)
+                return 0;
+
+        _cleanup_free_ struct local_address *addresses = NULL;
+        int n_addresses;
+        n_addresses = local_addresses(NULL, 0, port->ip_family, &addresses);
+        if (n_addresses < 0)
+                return log_error_errno(n_addresses, "Failed to determine local IP addresses: %m");
+
+        log_notice("NVMe-TCP: %s %s%sListening on %s (port %" PRIu16 ")",
+                   special_glyph(SPECIAL_GLYPH_ARROW_RIGHT),
+                   emoji_enabled() ? special_glyph(SPECIAL_GLYPH_WORLD) : "", emoji_enabled() ? " " : "",
+                   af_to_ipv4_ipv6(port->ip_family),
+                   port->portnr);
+
+        FOREACH_ARRAY(a, addresses, n_addresses) {
+                log_info("          %s Try for all devices: nvme connect-all -t tcp -a %s -s %" PRIu16,
+                         special_glyph(a >= addresses + (n_addresses - 1) ? SPECIAL_GLYPH_TREE_RIGHT : SPECIAL_GLYPH_TREE_BRANCH),
+                         IN_ADDR_TO_STRING(a->family, &a->address),
+                         port->portnr);
+        }
+
+        return 0;
+}
+
+typedef struct Context {
+        Hashmap *subsystems;
+        NvmePort *ipv4_port, *ipv6_port;
+
+        bool display_refresh_scheduled;
+} Context;
+
+static void device_hash_func(const struct stat *q, struct siphash *state) {
+        assert(q);
+
+        if (S_ISBLK(q->st_mode) || S_ISCHR(q->st_mode)) {
+                mode_t m = q->st_mode & S_IFMT;
+                siphash24_compress(&m, sizeof(m), state);
+                siphash24_compress(&q->st_rdev, sizeof(q->st_rdev), state);
+                return;
+        }
+
+        return inode_hash_func(q, state);
+}
+
+static int device_compare_func(const struct stat *a, const struct stat *b) {
+        int r;
+
+        assert(a);
+        assert(b);
+
+        r = CMP(a->st_mode & S_IFMT, b->st_mode & S_IFMT);
+        if (r != 0)
+                return r;
+
+        if (S_ISBLK(a->st_mode) || S_ISCHR(a->st_mode)) {
+                r = CMP(major(a->st_rdev), major(b->st_rdev));
+                if (r != 0)
+                        return r;
+
+                r = CMP(minor(a->st_rdev), minor(b->st_rdev));
+                if (r != 0)
+                        return r;
+
+                return 0;
+        }
+
+        return inode_compare_func(a, b);
+}
+
+DEFINE_PRIVATE_HASH_OPS_WITH_VALUE_DESTRUCTOR(
+                nvme_subsystem_hash_ops,
+                struct stat,
+                device_hash_func,
+                device_compare_func,
+                NvmeSubsystem,
+                nvme_subsystem_destroy);
+
+static void context_done(Context *c) {
+        assert(c);
+
+        c->ipv4_port = nvme_port_destroy(c->ipv4_port);
```

Comment:
Not a matter, but we usually not assign 0 or NULL in xyz_done().

Reply:
we sometimes do, and sometimes do not ;-)